### ISSUE TITLE
Generate report for complex test structure

### ISSUE THREAD

**Issue by cezarypiatek** [cezarypiatek/NScenario#18]
HI Cezary,
I trust you are fine.
Imagine you have a set of tests structured in folder and subfolders, which could provide together with the test class names semantic meaning to the report.
What would be the way to proceed to incorporate these structural elements in the md report?
Thanks !
paul.

_Originally posted by @paulvanbladel in https://github.com/cezarypiatek/NScenario/issues/10#issuecomment-1671313558_

**Comment by cezarypiatek** [comment]
@paulvanbladel  Could you propose the desired output? An example of such a report?

**Comment by paulvanbladel** [comment]
Well, it depends of course on the technical possibilities.
I think you use already the name of the test class, but I'm referring to elements that could
1. be derived from the file path of the current test class
2. be derived from a custom attribute provided above the current test class
In such a way a numbered structure (with md heading levels) is created around the test classes.
Imagine, you have a file structure above your tests with 3 levels.

Level A
  Level A1
    Level A1a
    Level A1b
    Level A1c
  Level A2
    Level A2a
    Level A2b
    Level A2c
Level B
  Level B1
    Level BAa
    ...
  Level B1
  Level B1
Level C
 Level C1
 Level C2

The level indicator typically would have a meaning specific to your project.

Makes sense?

**Comment by cezarypiatek** [comment]
I belive it's possible to retrieve the test file path and create derived info from that.

**Comment by paulvanbladel** [comment]
Yes, I recall something similar.
And how to write to the MD file with specific header formatting?

**Comment by cezarypiatek** [comment]
I can try to implement that but still not sure if I understand what do you need. What's your root problem? Can you provide an example markdown report that you expect as the output?

**Comment by paulvanbladel** [comment]
The current report nscenario functionality focusses on the markup and content of one specific test class.
If you want to put hierarchical structure in the report you have to rely on only that. But this gives a flat structure, in case you want to arrange the content like a book with chapters, sub chapters and so one, you need something more.
In my understanding, the file structure is the most natural base for providing that information.
A simple example could be a web app structured in feature and screens as expressed via the file folder structure.
If you could create headings in the report just for the feature and screen level, you can reserve more real estate for the actual textual elements that you expose (with the already in place functionality of nscenario) from the class and test level of the test.

Of course, an interesting addition to this would be having the possibility to create inside each folder one .md file that provides context information about the hierarchy level. E.g. some text giving general context about the feature or screen (as in my example). That short piece of text could be renderered then during the report generation;

So, what's the root problem. Having living documentation based on the structure of the tests. The points I add, would seriously improve the current possibilities.

**Comment by cezarypiatek** [comment]
Do you have a specific requirement to have that in markdown or the output could be an HTML file?
With HTML we have more possibilities in the data presentation area?

**Comment by paulvanbladel** [comment]
Not necessarily, what i would think of is just a different ## ### ### title level corresponding with the level of the folder structure.
But, in case html generation would be easier, i don't see any problem.

**Comment by cezarypiatek** [comment]
I've started working on a better test report. What do you think?

![image](https://github.com/cezarypiatek/NScenario/assets/7759991/ac4c53ab-7925-4b26-aabe-59beb28f13fa)

**Comment by paulvanbladel** [comment]
Love it.

**Comment by cezarypiatek** [comment]
The option to generate a new report should be available in NScenario v5.1.
Just add the following line in your global teardown

```cs
TestScenarioFactory.GetAllExecutedScenarios().SaveAsReport("AllReports.html");
```

Any feedback is welcome.
